tests/unit/plugins/test_event_consumer.py

```diff
@@ -15,23 +15,549 @@
+                                            side_effect, expected):
+    """Validate & return schema of a pubsub message's data."""
+    config = {'subscription': 'foo'}
+    parser = parse.MessageParser()
+    consumer = event_consumer.GPSEventConsumer(
+        config, None, validator, parser, None, None, None)
+
+    validator.validate.side_effect = side_effect
+    data = json.loads(pubsub_msg.data)
+    schema = consumer._get_and_validate_pubsub_msg_schema(data)
+
+    calls = [mocker.call(data, 'schema1')]
+    if not side_effect[0]:
+        calls.append(mocker.call(data, 'schema2'))
+    validator.validate.assert_has_calls(calls)
+
+    assert expected == schema
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize('init_phase', (None, 'emo'))
+async def test_update_phase(init_phase, mocker, consumer, pubsub_msg):
+    """Phase of a GEventMessage instance is updated."""
+    event_msg = event_consumer.GEventMessage(pubsub_msg, {}, phase=init_phase)
+
+    assert init_phase == event_msg.phase  # sanity check
+    with mock.patch(DATETIME_PATCH, conftest.MockDatetime):
+        await consumer.update_phase(event_msg)
+
+    assert 'consume' == event_msg.phase
+    exp_entry = {
+        'timestamp': '2018-01-01T11:30:00.000000Z',
+        'plugin': 'consume',
+        'message': f'Updated phase from "{init_phase}" to "consume".',
+    }
+    assert exp_entry in event_msg.history_log
+
+
+@pytest.mark.asyncio
+async def test_cleanup(mocker, consumer, caplog, pubsub_msg):
+    """Consumer acks Pub/Sub message."""
+    event_msg = event_consumer.GEventMessage(pubsub_msg, {})
+    mock_event = mocker.Mock(threading.Event)
+    mock_thread = mocker.Mock(event_consumer._GPSThread)
+    ack_id = 'projects/a-project/subscriptions/test-sub:1413'
+    consumer._threads[ack_id] = (mock_event, mock_thread)
+    with mock.patch(DATETIME_PATCH, conftest.MockDatetime):
+        await consumer.cleanup(event_msg)
+
+    pubsub_msg.ack.assert_called_once_with()
+
+    exp_entry = {
+        'timestamp': '2018-01-01T11:30:00.000000Z',
+        'plugin': 'consume',
+        'message': 'Acknowledged message in Pub/Sub.',
+    }
+    assert exp_entry in event_msg.history_log
+    mock_event.set.assert_called_once_with()
+    mock_thread.stop.assert_called_once_with()
+    assert 2 == len(caplog.records)
+
+
+#####
+# _GPSThread tests
+#####
+def test_create_gpsthread(thread, event):
+    gps_thread = event_consumer._GPSThread(event)
+
+    assert not gps_thread.loop
+    assert not gps_thread.thread_id
+    assert event is gps_thread.event
+
+
+class CustomLoop(asyncio.BaseEventLoop):
+    def run_forever(self):
+        # so we don't _actually_ run forever
+        pass
+
+    def stop(self):
+        pass
+
+    def close(self):
+        pass
+
+    def call_soon_threadsafe(self, *args, **kwargs):
+        pass
+
+
+@pytest.fixture
+def custom_event_loop():
+    loop = CustomLoop()
+    yield loop
```

Comment:
Is using yield->loop.close() necessary if `close` is a noop?

Reply:
lolnope.